File: Programmers/Before2022/PG_문자열압축/StringCompression.java
import java.util.ArrayList;

public class StringCompression {
    public static void main(String[] args) {
        System.out.println(Solution("a"));
    }

    private static int Solution(String s) {
        int answer = 1;

        for (int n = 1; n <= s.length() / 2; n++) {
            String now = "";
            int count = 0;

            ArrayList<String> strings = new ArrayList<>();
            int start = 0;
            int end = start + n;
            while(end < s.length()) {
                strings.add(s.substring(start, end));
                start = end;
                end = start + n;
            }
            strings.add(s.substring(start));

            StringBuilder sb = new StringBuilder();

            for(String currentStr : strings) {
                if (now.equals("")) {
                    now = currentStr;
                    count++;
                    continue;
                }

                if (now.equals(currentStr)) {
                    count++;
                } else {
                    if (count <= 1) {
                        sb.append(now);
                    } else {
                        sb.append(count).append(now);
                    }
                    now = currentStr;
                    count = 1;
                }
            }

            if (count <= 1) {
                sb.append(now);
            } else {
                sb.append(count).append(now);
            }

            if (answer == 1) {
                answer = sb.length();
            } else {
                answer = Math.min(answer, sb.length());
            }
        }

        return answer;
    }
}

File: Programmers/Before2022/PG_문자열압축/StringCompression.py
def string_comp(s):
    answer = 1

    for n in range(1, len(s) // 2 + 1):
        now = ""
        count = 0
        strings = []

        start = 0
        end = start + n
        while end < len(s):
            strings.append(s[start:end])
            start = end
            end = start + n
        strings.append(s[start:])

        temp = ""

        for cs in strings:
            if now == "":
                now = cs
                count += 1
            else:
                if now == cs:
                    count += 1
                else:
                    if count <= 1:
                        temp += "{}".format(now)
                    else:
                        temp += "{}{}".format(count, now)
                    now = cs
                    count = 1

        if count <= 1:
            temp += "{}".format(now)
        else:
            temp += "{}{}".format(count, now)

        if answer == 1:
            answer = len(temp)
        else:
            answer = min(answer, len(temp))

    return answer


result = string_comp("ababcdcdababcdcd")
print(result)
